### ISSUE TITLE
Minimal Working Example with Non-root User (Jovyan) and Single User Spawn

### ISSUE THREAD

**Issue by ColinConwell** [ml-tooling/ml-workspace#11]
MLWorkspace strikes me as a phenomenal tool with broad applicability. I'm wondering if you might be able to provide an example of how to fit it into a standard multi-user jupyterhub + dockerspawner, where a 'start-singleuser' command spawns an image in which the primary user is not root, but jupyter's standard jovyan, and permissions are fixed accordingly. I've been attempting this, but it's proving a bit elusive to me. 

Basically, is there a way of combining jupyter's base-notebook stack with mlworkspace to create a single-user compatible image that could be spawned from jupyterhub?

(MLHub seems like a natural step in this direction, but this would be for users who are interested in integrating MLWorkspace into their existing workflows and who might be gunshy about having single users have root permissions).

My thanks in advance.

**Comment by lukasmasuch** [comment]
Hey Colin, thanks a lot for your kind words and the feedback!

To a certain degree, the Workspace is already compatible with JupyterHub (e.g., it is using the same start-scripts as docker-stacks images). However, since JupyterHub is not able to perform a universal routing (proxying) of any kind of HTTP/SSH request, a lot of the additional functionality of the ML Workspace such as SSH access or accessing some tools (VS Code, VNC, Ungit, ...) will not work. This is one reason why we created ML Hub on top of JupyterHub to add these universal routing capabilities.

As with root permission, we have just added a small section about this topic to our [FAQ](https://github.com/ml-tooling/ml-workspace#faq). So, currently, it is not supported, but we are planning to add this capability. Since we do not see it as supercritical from a security perspective (as explained in the FAQ), this might take some time until we can get this implemented. We are planning a major refactoring (updating to Python 3.7, Ubuntu 18, adding CI tests...) until end of this year, and adding a non-root user might be one additional goal for this task.

**Comment by sim-san** [comment]
Can the major refactoring be expected this year ?

**Comment by jmuchovej** [comment]
@LukasMasuch @ColinConwell So, maybe, instead of adding a user within the container, it might make sense to do a UID:GID mapping to the user and their group? I don't believe it's a huge security risk to mount `/etc/passwd:ro` within the container.

Proposing this approach, instead, mostly because a new user within the container will always have PID/GID = 1000, unless otherwise specified, and Docker will perform that mapping within the container. In the case that you're not 1000:1000 on the host system, you'll still run into the same problem as having files owned root:root. _At least that's based on a rather limited understanding of both Docker and Linux, in the context of containers._

**Comment by usmcamp0811** [comment]
What can I do to help speed this along? I've got an environment with a bunch of NFS mounts that won't let any user that is not in our AD to access them. So without the ability to set the UID:GID of a user they won't be able to access any files. 

I was playing around with things to see if just creating a new user with `root` privileges that would run the `supervisord` stuff, and I got `jupyter lab` to start but it started as `root`.  

Is there / or can we start a list of what gets started and whether or not it needs root run? I don't mind doing the leg work, there is just a lot to digest here. 

Or is there another thread this is being tracked on? 

What can I do to help.. just point me in the right direction and give me some marching orders..

**Comment by jmuchovej** [comment]
@usmcamp0811 So, I can't speak to ML Tooling's progress; but I had planned to make some headway on this in the next month or so (have a few paper deadlines coming up so those take priority for me). 

I've worked with JupyterHub in the past and have deployed it on compute clusters as well.

There were a few modifications I was planning to make:
1. Attempt a migration from the current `Dockerfile` to  an [`s6-overlay`][s6]. This is mostly to allow for insertion of tasks to run on startup in a slightly cleaner and more understandable way.
1. Add a non-root (e.g. `abc`) who takes on the `$PUID:$PGID` passed into the container via environment variables. (This is straight out of the inner workings of `linuxserver` containers.)

I haven't picked apart the container, but I don't believe many (if any) of these services need to be run as root. In the event there is one, I imagine using something like `s6` might be a way around it.

I'm not sure if you're looking to integrate Jupyter**Hub**, but if so, adding support for this is something I've been looking in to (though I've been backlogged with work and research since posting). https://github.com/ml-tooling/ml-hub/issues/5

I'm not sure what your background is, but I'd be down to split time across this.

[s6]: https://github.com/just-containers/s6-overlay

**Comment by lukasmasuch** [comment]
Hey, awesome to see that you are spending your time on this and want to improve it. We (= ml-tooling maintainers -> @raethlein @janukowitsch ) are quite heavily involved in some official Corona-related projects that are somewhat time-pressuring and, honestly, we are not able to focus on ml-tooling right now :( But we just wanted to let you know that we really really appreciate your effort here and will shift focus back to it as soon as our other tasks slowed down. So we are happy to jump back in the discussion more actively later on or merge any pull-requests / progress that happened so far.

**Comment by usmcamp0811** [comment]
@ionlights I'm kind of a "jack of all trades, master of very little" type of guy. I've never heard of `s6-overlay`, but that looks pretty interesting. I'll try and do some hello world stuff with it over the next couple of days to get familiar with it and hopefully I can help you out. 

For simplicity sake I like the way `linuxserver` handles user id stuff nicely.. though I did make a simple Python script the other day to do a little more complete cloning of a user, and I don't mind offering it up. The script can be found [here](https://raw.githubusercontent.com/usmcamp0811/campfiles/master/id-clone.py). It essentially takes in the output of `$(id <user>)` and will create that user, and all the groups the user belongs to. If there are conflicting groups in the system it will changes their `ID` to be something else.

**Comment by jmuchovej** [comment]
@usmcamp0811 The script looks pretty cool. 🙂 I have a few questions, though:
- **How are you envisioning this being used?**
- **Are you envisioning each user group being created dynamically?** For that to make it's way into the container, I believe that would need to be added as build?
- **What would be a case where a single `MLWorkspace` container needs multiple groups?** (I can't think of an example, hence the question.)
  - I see a case where `MLHub` might need that, but I think mounting `/etc/passwd:ro` might be a feasible solution since the IDs are needed, if that makes sense?

As for `s6-overlay` – yea, it's a great project; definitely makes it a bit easier to parse what the startup tasks are.

I think, if `s6-overlay` is setup, that would make the process of `chown`-ing files much easier and less "hacky" as we can just set the sticky-bit on the `/workspace` folder within the container.

Once I can take a proper crack at `SystemUserSpawner` (mentioned in the other thread), I think that should resolve [almost] everything you mentioned above – and what other users have also seemed to want.

`SystemUserSpawner` uses the `/etc/passwd` database to authenticate and spawn individual Jupyter sessions. In this case, running `MLHub` would spawn `MLWorkspaces` using `SystemUserSpawner`.

**Comment by usmcamp0811** [comment]
Thanks,

> 
**How are you envisioning this being used?**
- I've got an environment that's pretty locked down, and all users are managed by an Active Directory. All home directories are on a central NFS server and require correct user ID and groups to access the files.
>
**Are you envisioning each user group being created dynamically?** 
- I've got a container I have for work that I used it on. It doesn't have to be done at build time, just run in an entrypoint script. It might have one little bug in it as I ran it at work and for some reason the primary user of my container didn't get its UID changed, but all the other groups and what not were good to go. 
>
**What would be a case where a single MLWorkspace container needs multiple groups?** 
- we probably could use `/etc/passwd` **but** I can't in my environment because all groups aren't stored in it in my environment. Not 100% sure how or why.. but pretty sure its got something to do with Active Directory. But there are some cases where my users would need multiple groups to be able to access files. Like I said we have a very locked down system...  :-( 


 I haven't had a chance yet to play with `s6-overlay` but hope to this weekend.
